### PR TITLE
Migrate host/v5 metadata payload to component

### PR DESCRIPTION
**Pull request by hush-hush**
### What does this PR do?

The "host" payload, previously names "v5", is now a component. The core of the payload lives in comp/metadata/host/utils until the rest of the code base using it is migrated to component too.

### Describe how to test/QA your changes

Test on windows, OSX and Linux that the v5 metadata payload is the same than in the previous version of the agent.

For process: fomr (https://github.com/DataDog/datadog-agent/pull/19463#pullrequestreview-1668806558)

This PR updates the host metadata payload which is used in the process agent to display the hostname in the status check, so the expected behavior here is that when we run `sudo -u dd-agent /opt/datadog-agent/embedded/bin/process-agent status`, the hostname still shows up correctly in the process agent status output e.g.
```
  Enabled Checks: [process rtprocess connections]
  Allocated Memory: 18,253,560 bytes
  Hostname: kkhor-vagrant
```

This is a pretty simple test so if your team wants to do the agent QA for this that would be fine, otherwise you can add the `team/processes` tag so it shows up on our board during agent QA

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by ogaca-dd — COMMENTED** [review]
LGTM on the component framework side.
Note that I didn't check that the code behaves the same as before

**Review by pgimalac — APPROVED** [review]
I didn't catch any big issue, mostly just a few things removed or added without understanding where they came from.
There are so many changes all over that I think extensive QA will be the best way to catch a bug

File: cmd/dogstatsd/subcommands/start/command.go [hunk 1]
```diff
@@ -102,10 +107,35 @@ func RunDogstatsdFct(cliParams *CLIParams, defaultConfPath string, defaultLogFil
 		dogstatsd.Bundle,
 		forwarder.Bundle,
 		fx.Provide(defaultforwarder.NewParams),
+		// injecting the aggregator demultiplexer to FX until we migrate it to a proper component. This allows
+		// other already migrated components to request it.
+		fx.Provide(func(config config.Component, log log.Component, sharedForwarder defaultforwarder.Component) (*aggregator.AgentDemultiplexer, error) {
```
Comment (vickenty) [inline cmd/dogstatsd/subcommands/start/command.go:112]: Would it make sense to extract this into a named function, to keep the component and bundle list manageable?

File: cmd/dogstatsd/subcommands/start/command.go [hunk 2]
```diff
@@ -191,27 +222,9 @@ func RunAgent(ctx context.Context, cliParams *CLIParams, config config.Component
...
-		log.Warnf("Error getting hostname: %s", err)
-		hname = ""
-	}
-	log.Debugf("Using hostname: %s", hname)
-	demux := aggregator.InitAndStartAgentDemultiplexer(log, forwarder, opts, hname)
-	demux.AddAgentStartupTelemetry(version.AgentVersion)
-
-	// setup the metadata collector
-	components.MetaScheduler = metadata.NewScheduler(demux) //nolint:staticcheck
-	if err = metadata.SetupMetadataCollection(components.MetaScheduler, []string{"host"}); err != nil {
```
Comment (vickenty) [inline cmd/dogstatsd/subcommands/start/command.go:209]: This call seems to be absent from the new code, is it expected?

File: pkg/util/cloudproviders/network.go [hunk 3]
```diff
@@ -16,38 +16,38 @@ import (
 	"github.com/DataDog/datadog-agent/pkg/util/log"
 )
 
+const (
+	networkIDCacheKey = "network.id"
```
[the remark this answers is not in the document]
Comment (hush-hush) [reply pkg/util/cloudproviders/network.go:20]: No it's a typo. Still works but I'll fix for consistency.

File: comp/metadata/host/utils/host.go [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+	return p
+}
+
+// GetFromCache returns the payload from the cache if it exists, otherwise it creates it.
+// The metadata reporting should always grab it fresh. Any other uses, e.g. status, should use this
+func GetFromCache(ctx context.Context, conf config.ConfigReader) *Payload {
+	data, found := cache.Cache.Get(hostCacheKey)
+	if !found {
+		return GetPayload(ctx, conf)
+	}
```
[the remark this answers is not in the document]
Comment (hush-hush) [reply comp/metadata/host/utils/host.go:200]: It should but it didn't in the past so I migrated it this way. Once the migration is done we can start introducing breaking change in behavior. But for now I'd like to keep the PR simpler.

File: comp/metadata/host/utils/host_windows.go [hunk 5]
```diff
@@ -65,16 +71,53 @@ func GetInformation() *InfoStat {
 			info.KernelArch = runtime.GOARCH
 
 			pi := platform.CollectInfo()
-			info.Platform = pi.OS.ValueOrDefault()
-			info.PlatformFamily = pi.OS.ValueOrDefault()
+			info.Platform, _ = pi.OS.Value()
+			info.PlatformFamily, _ = pi.OS.Value()
```
[the remark this answers is not in the document]
Comment (hush-hush) [reply comp/metadata/host/utils/host_windows.go:75]: I don't know

File: cmd/dogstatsd/subcommands/start/command.go (hunk 2 above)
Comment (hush-hush) [reply cmd/dogstatsd/subcommands/start/command.go:209]: Yes, the `host` metadata is now a component and no longer require all the init and setup phase.

File: cmd/agent/subcommands/run/command.go [hunk 6]
```diff
@@ -479,13 +482,6 @@ func startAgent(
 	}
 	log.Infof("Hostname is: %s", hostnameDetected)
 
-	// HACK: init host metadata module (CPU) early to avoid any
-	//       COM threading model conflict with the python checks
-	err = host.InitHostMetadata()
```
[the remark this answers is not in the document]
Comment (hush-hush) [reply cmd/agent/subcommands/run/command.go:424]: This was already useless but never removed. In the past when we stopped using `gopsutil` on windows this was left behind. As part of the work on the metadata and its cleanup I realized we no longer need this for a while and removed it.

File: comp/metadata/host/utils/host_windows.go [hunk 7]
```diff
@@ -3,27 +3,33 @@
 // This product includes software developed at Datadog (https://www.datadoghq.com/).
 // Copyright 2016-present Datadog, Inc.
 
-// Portions of this code are taken from the gopsutil project
```
[the remark this answers is not in the document]
Comment (hush-hush) [reply comp/metadata/host/utils/host_windows.go:6]: Probably not, I'll add it back.

File: comp/metadata/resources/resources.go [hunk 8]
```diff
@@ -36,6 +36,15 @@ type resources struct {
...
 	fx.In
 
+	// Resources is enabled by default for most binaries. But even for binaries where we don't want to send the
+	// 'resources' paylod, like dogstatsd, we still need the resources Component. This is because the resources data
+	// is embedded in other metadata payload like 'host'. This means that even if resources is disabld it might be
+	// require in the build to the `Get` method is available.
+	//
+	// This is why we have a Params struct for resources. It's `optional` so most of the binaries don't have to
+	// supply a Params struct but only need to import the metadata.Bundle.
+	Params *Params `optional:"true"`
```
[the remark this answers is not in the document]
Comment (hush-hush) [reply comp/metadata/resources/resources.go:46]: Here we're not depending on a component or the result of some computation. Their are no error that can be hidden in term of dependencies.

We want to enable `resources` by default while letting code that what's to to disable it. I think having multiple structs, methods and overwrites is a lot of code to maintain for a pointer. And, here, `optional:"true"` is exactly what we want.

This also has no consequences outside the internal of the `resources` provider.

**Review by vickenty — APPROVED** [review]

**Review by kkhor-datadog — COMMENTED** [review]
Mind updating the QA instructions for the process agent functionality? It looks like this PR updates the host metadata payload which is used in the process agent to display the hostname in the status check, so the expected behavior here is that when we run `sudo -u dd-agent /opt/datadog-agent/embedded/bin/process-agent status`, the hostname still shows up correctly in the process agent status output e.g.
```
  Enabled Checks: [process rtprocess connections]
  Allocated Memory: 18,253,560 bytes
  Hostname: kkhor-vagrant
```

This is a pretty simple test so if your team wants to do the agent QA for this that would be fine, otherwise you can add the `team/processes` tag so it shows up on our board during agent QA

**Review by ogaca-dd — APPROVED** [review]

**Review by liustanley — APPROVED** [review]

**Review by kkhor-datadog — APPROVED** [review]

**Pull request merged**